aws-rds-cfn-common/src/main/java/software/amazon/rds/common/util/DriftDetector.java

```diff
@@ -0,0 +1,252 @@
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.BooleanUtils;
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.Schema;
+
+import com.api.jsonata4java.expressions.EvaluateException;
+import com.api.jsonata4java.expressions.Expressions;
+import com.api.jsonata4java.expressions.ParseException;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import software.amazon.cloudformation.resource.ResourceTypeSchema;
+
+public class DriftDetector {
+
+    private static final String INSERTION_ORDER = "insertionOrder";
+    private static final String PROPERTIES_ROOT = "/properties";
+    private static final String PROPERTY_OR_SPLIT_REGEX = "\\s+\\$OR\\s+";
+    private static final String PROPERTY_PATH_SEPARATOR = "/";
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    private static final Map<String, Mutation> NO_DRIFT = Collections.emptyMap();
+
+    private final ResourceTypeSchema schema;
+    private final Set<String> writeOnlyLookup;
+
+    public DriftDetector(final ResourceTypeSchema schema) {
+        this.schema = schema;
+        this.writeOnlyLookup = new HashSet<>(
+                Optional.ofNullable(schema.getWriteOnlyPropertiesAsStrings())
+                        .orElse(Collections.emptyList())
+        );
+    }
+
+    private static boolean isTransformableProperty(
+            final ResourceTypeSchema schema,
+            final String propertyName
+    ) throws DriftDetectorRuntimeException {
+        return schema.getPropertyTransform().containsKey(propertyName);
+    }
+
+    private static Object castToPrimitive(final JsonNode node, final Class<?> klass) {
+        if (Integer.class.equals(klass)) {
+            return node.asInt();
+        } else if (Boolean.class.equals(klass)) {
+            return node.asBoolean();
+        } else if (String.class.equals(klass)) {
+            return node.asText();
+        }
+
+        throw new DriftDetectorRuntimeException("Unexpected cast input class: " + klass.getName());
+    }
+
+    private List<Object> getTransformations(
+            final Object in,
+            final String path,
+            final JsonNode rootNode
+    ) {
+        final List<Object> result = new ArrayList<>();
+        final String propTxs = schema.getPropertyTransform().get(path);
+        if (propTxs != null) {
+            for (final String tx : propTxs.split(PROPERTY_OR_SPLIT_REGEX)) {
+                try {
+                    final Expressions expr = Expressions.parse(tx);
+                    final JsonNode txNode = expr.evaluate(rootNode);
+                    result.add(castToPrimitive(txNode, in.getClass()));
```

Comment:
If I understand correctly, JSONata also supports Array and object transformations, e.g.: https://docs.jsonata.org/array-functions and https://docs.jsonata.org/object-functions

It might be worth indicating in the exception message that the transform feature is not supported for those types yet.

Reply:
Currently it would throw a DriftDetectorRuntimeException with the following error message: `Unexpected cast input class: List<...>`. Shall we go any further than that?